.github/actions/yarn-nm-install/action.yml

```diff
@@ -0,0 +1,88 @@
+#                                                                                      #
+#      - name: 📥 Monorepo install                                                     #
+#        uses: ./.github/actions/yarn-nm-install                                       #
+#        with:                                                                         #
+#          enable-corepack: false # (default)                                          #
+#          cache-install-state: false # (default)                                      #
+#          cache-node-modules: false # (default)                                       #
+#                                                                                      #
+# Reference:                                                                           #
+#   - latest: https://gist.github.com/belgattitude/042f9caf10d029badbde6cf9d43e400a    #
+########################################################################################
+
+name: 'Monorepo install (yarn)'
+description: 'Run yarn install with node_modules linker and cache enabled'
+inputs:
+  enable-corepack:
+    description: 'Enable corepack'
+    required: false
+    default: 'false'
+  cache-node-modules:
+    description: 'Cache node_modules, might speed up link step (invalidated lock/os/node-version/branch)'
+    required: false
+    default: 'false'
+  cache-install-state:
+    description: 'Cache yarn install state, might speed up resolution step when node-modules cache is activated (invalidated lock/os/node-version/branch)'
+    required: false
+    default: 'false'
+
+runs:
+  using: 'composite'
+
+  steps:
+    - name: ⚙️ Enable Corepack
+      if: ${{ inputs.enable-corepack }} == 'true'
+      shell: bash
+      run: corepack enable
+
+    - name: ⚙️ Expose yarn config as "$GITHUB_OUTPUT"
+      id: yarn-config
+      shell: bash
+      env:
+        YARN_ENABLE_GLOBAL_CACHE: "false"
+      run: |
+        echo "CACHE_FOLDER=$(yarn config get cacheFolder)" >> $GITHUB_OUTPUT
+        echo "CURRENT_NODE_VERSION="node-$(node --version | cut -d . -f 1  | sed 's/[^0-9]*//g')".x" >> $GITHUB_OUTPUT
+        echo "CURRENT_BRANCH=$(echo ${GITHUB_REF#refs/heads/} | sed -r 's,/,-,g')" >> $GITHUB_OUTPUT
+
+    - name: ♻️ Restore yarn cache
+      uses: actions/cache@v3
+      id: yarn-download-cache
+      with:
+        path: ${{ steps.yarn-config.outputs.CACHE_FOLDER }}
+        key: yarn-download-cache-${{ hashFiles('yarn.lock', '.yarnrc.yml') }}
+        restore-keys: |
+          yarn-download-cache-
+
+    - name: ♻️ Restore node_modules
+      if: inputs.cache-node-modules == 'true'
+      id: yarn-nm-cache
+      uses: actions/cache@v3
+      with:
+        path: '**/node_modules'
+        key: yarn-nm-cache-${{ runner.os }}-${{ steps.yarn-config.outputs.CURRENT_NODE_VERSION }}-${{ steps.yarn-config.outputs.CURRENT_BRANCH }}-${{ hashFiles('yarn.lock', '.yarnrc.yml') }}
```

Comment:
Just wondering if that is a bit over-optimized. Can we assume that `CURRENT_NODE_VERSION` and `hashFiles` is enough? Why do you think `CURRENT_BRANCH` should be part?

Reply:
Note that based on the strapi repo.  The node-modules folder isn't cached if not requested otherwise. False by default.

This is based on a previous benchmark made in this branch. 

The reason node version is added is to be future proof. Imagine someone adds a library that will run a node-gyp postinstall build to create a binary.... for example sharp when vips-dev isn't installed.... there's other examples in the wild,  I've chosen sharp cause you have it. It saves the compiled binary in the node modules folder and... 💥 

But the most interesting reason to keep this... keep inline with the original gist that comes with doc.

In a next iteration,  I'll plan to add more speed up. I prefer to restart from a situation I know well

Reply:
Gu-stav is right @belgattitude I think that CURRENT_BRANCH is not needed since nodeversion and packages.yml already solves this what would mean less cold starts

Reply:
What @Boegie19 said :) `CURRENT_NODE_VERSION` looks useful to me - just confused about `CURRENT_BRANCH` :)

Reply:
Ah yes, haven't explained this...

The idea is that yarn/pnpm works in multiple phases: 

- Resolution - seems affected by presence of install_state
- Fetch - only affected by zip cache presence
- Link - seems affected by presence of node_modules/install_state

So why it's there ? 

Over time in a repo... it's always nice to re-asses cost/benefit perf between the two possible strategies

- Just cache *.zip 
- Cache *.zip + install-state + node_modules.

And that depends mostly on the packages installed in the repo (ie you install esbuild, vite, vitest at some point and the balance is different)

In the strapi repo

- The strategy that makes sense is to just save *.zip
- As in one of my repos, it's better to save everything

![image](https://user-images.githubusercontent.com/259798/236443796-5c016d05-7f33-4ee0-8bc9-8b44469ac617.png)

See https://github.com/belgattitude/nextjs-monorepo-example/actions/runs/4888683373/jobs/8726574932 (brings a 1 minute with strategy 1 install to 25seconds install) - not totally correct but that's the idea...

The drawback node_modules/install-state cannot leak between branches (long to explain). The best for now is to add the branch to the key.  

Yes it's over-optimization in the current strapi repo... and it might look hard-to-read :) With another repo this optim is valid. (mitgh make sense in the future)

But the main idea is to not diverge too much from the reference gist: https://gist.github.com/belgattitude/042f9caf10d029badbde6cf9d43e400a

This is optimization after-all, better to have a reference doc

Reply:
I see you point. What do you think would be the danger in sharing a cache across branches? We don't bundle/ inline any secrets, so I don't see a problem there - but maybe for caches (thinking e.g. of eslint / prettier) that might be shared and shouldn't?

Reply:
Link step and whatever will happen there + tons of things... but too long to explain today. As it's not enabled... let's keep it as close to reference gist if you agree 

Wdyt ? Or you want more details ?

Reply:
Take your time - I'd like to know before we merge what the reason for that is, because otherwise we won't be able to get rid of it again.